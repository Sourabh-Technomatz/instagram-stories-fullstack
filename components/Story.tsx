import React from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes } from '@fortawesome/free-solid-svg-icons';

interface StoryProps {
  src: string;
  name: string;
  onPrev: () => void;
  onNext: () => void;
  progress: number;
  onClose: () => void;
}

const Story: React.FC<StoryProps> = ({ src, name, onPrev, onNext, progress, onClose }) => {
  return (
    <div className="flex items-center justify-center h-full w-full bg-black relative">
      <AnimatePresence mode="wait">
        <motion.img
          key={src}
          src={src}
          alt="Story"
          initial={{ opacity: 0, x: 300 }}
          animate={{ opacity: 1, x: 0 }}
          exit={{ opacity: 0, x: -300 }}
          transition={{ duration: 0.5 }}
          className="max-h-full max-w-full object-cover"
        />
      </AnimatePresence>
      <div
        className="absolute left-0 top-0 h-full w-1/2 cursor-pointer"
        onClick={onPrev}
        data-testid="left-half"
      />
      <div
        className="absolute right-0 top-0 h-full w-1/2 cursor-pointer"
        onClick={onNext}
        data-testid="right-half"
      />
      <div
        className="absolute top-0 left-0 h-1 bg-red-500"
        style={{ width: `${progress}%`, transition: 'width 0.1s linear' }}
        data-testid="progress-bar"
      />
      <div className="absolute top-2 left-4 right-4 flex items-center justify-between">
        <div className="text-white text-lg bg-transparent px-2 py-1 rounded">
          {name}
        </div>
        <div className="p-2 cursor-pointer" onClick={onClose} data-testid="close-button">
          <FontAwesomeIcon icon={faTimes} className="text-white text-2xl" />
        </div>
      </div>
    </div>
  );
};

export default Story;
