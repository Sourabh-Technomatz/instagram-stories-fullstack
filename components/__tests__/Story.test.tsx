import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Story from '../Story';

jest.mock('@fortawesome/react-fontawesome', () => ({
  FontAwesomeIcon: (props: any) => <i {...props}></i>,
}));

describe('Story component', () => {
  const mockProps = {
    src: 'test-image.jpg',
    name: 'Test Story',
    onPrev: jest.fn(),
    onNext: jest.fn(),
    progress: 50,
    onClose: jest.fn(),
  };

  it('should render the component correctly and match snapshot', () => {
    const { asFragment } = render(<Story {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display the correct image', () => {
    render(<Story {...mockProps} />);
    const img = screen.getByAltText('Story');
    expect(img).toHaveAttribute('src', 'test-image.jpg');
  });

  it('should display the correct name', () => {
    render(<Story {...mockProps} />);
    const nameElement = screen.getByText('Test Story');
    expect(nameElement).toBeInTheDocument();
  });

  it('should set the progress bar width based on progress prop', () => {
    render(<Story {...mockProps} />);
    const progressBar = screen.getByTestId('progress-bar');
    expect(progressBar).toHaveStyle('width: 50%');
  });

  it('should call onPrev when the left half is clicked', () => {
    render(<Story {...mockProps} />);
    const leftHalf = screen.getByTestId('left-half');
    fireEvent.click(leftHalf);
    expect(mockProps.onPrev).toHaveBeenCalled();
  });

  it('should call onNext when the right half is clicked', () => {
    render(<Story {...mockProps} />);
    const rightHalf = screen.getByTestId('right-half');
    fireEvent.click(rightHalf);
    expect(mockProps.onNext).toHaveBeenCalled();
  });

  it('should call onClose when the close button is clicked', () => {
    render(<Story {...mockProps} />);
    const closeButton = screen.getByTestId('close-button');
    fireEvent.click(closeButton);
    expect(mockProps.onClose).toHaveBeenCalled();
  });
});
