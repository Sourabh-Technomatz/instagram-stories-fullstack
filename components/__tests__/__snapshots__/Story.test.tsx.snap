// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story component should render the component correctly and match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center h-full w-full bg-black relative"
  >
    <img
      alt="Story"
      class="max-h-full max-w-full object-cover"
      src="test-image.jpg"
      style="opacity: 0; will-change: opacity,transform; transform: translateX(300px);"
    />
    <div
      class="absolute left-0 top-0 h-full w-1/2 cursor-pointer"
      data-testid="left-half"
    />
    <div
      class="absolute right-0 top-0 h-full w-1/2 cursor-pointer"
      data-testid="right-half"
    />
    <div
      class="absolute top-0 left-0 h-1 bg-red-500"
      data-testid="progress-bar"
      style="width: 50%; transition: width 0.1s linear;"
    />
    <div
      class="absolute top-2 left-4 right-4 flex items-center justify-between"
    >
      <div
        class="text-white text-lg bg-transparent px-2 py-1 rounded"
      >
        Test Story
      </div>
      <div
        class="p-2 cursor-pointer"
        data-testid="close-button"
      >
        <i
          class="text-white text-2xl"
          icon="[object Object]"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
